src/rust/inetstack/mod.rs

```diff
@@ -273,27 +287,21 @@ impl InetStack {
         trace!("accept(): {:?}", qd);
 
         // Search for target queue descriptor.
-        match self.file_table.get(qd) {
-            // Found, check if it concerns a TCP socket.
-            Some(qtype) => match QType::try_from(qtype) {
-                // It does, so allocate a new queue descriptor and issue accept operation.
-                Ok(QType::TcpSocket) => {
-                    let new_qd: QDesc = self.file_table.alloc(QType::TcpSocket.into());
-                    let future: FutureOperation = FutureOperation::from(self.ipv4.tcp.do_accept(qd, new_qd));
-                    let handle: SchedulerHandle = match self.scheduler.insert(future) {
-                        Some(handle) => handle,
-                        None => {
-                            self.file_table.free(new_qd);
-                            return Err(Fail::new(libc::EAGAIN, "cannot schedule co-routine"));
-                        },
-                    };
-                    Ok(handle.into_raw().into())
-                },
-                // This queue descriptor does not concern a TCP socket.
-                _ => Err(Fail::new(libc::EINVAL, "invalid queue type")),
+        match self.lookup_qtype(&qd) {
+            Some(QType::TcpSocket) => {
+                let future: FutureOperation = FutureOperation::from(self.ipv4.tcp.do_accept(qd));
+                let handle: SchedulerHandle = match self.scheduler.insert(future) {
+                    Some(handle) => handle,
+                    None => {
+                        return Err(Fail::new(libc::EAGAIN, "cannot schedule co-routine"));
+                    },
+                };
+                Ok(handle.into_raw().into())
```

Comment:
I've seen that you have transferred the responsibility of queue descriptor allocation to the TCP protocol. This may look cleaner, because it removes the required cleanup code when the future completes. However, it worsens the boundaries between the network protocol layer and the sockets api layer. I suggest that we have a discussion on this.

Reply:
Yes, I agree. I will eventually replace this with a lower runtime layer that is shared across the protocols and libOSes, which manages the table across libOSes and allocates queue descriptors for everyone. I'm just pushing this down for now but it won't remain here.

Reply:
issue #495